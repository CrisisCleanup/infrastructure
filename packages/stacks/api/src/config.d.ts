import type {ApiStackConfig, ApiStackConfigInput} from './schema'

declare module '@crisiscleanup/config' {
  export interface CrisisCleanupConfigInput {
    apiStack?: ApiStackConfig
  }
  export interface CrisisCleanupConfig {
    apiStack?: ApiStackConfig
  }
}
