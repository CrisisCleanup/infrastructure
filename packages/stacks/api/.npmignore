# ~~ Generated by projen. To modify, edit .projenrc.js and run "npx projen".
/.projen/
/.prettierignore
/.prettierrc.json
/test/
/src/
!/dist/
!/dist/**/*.js
!/dist/**/*.d.ts
dist
/.github/
/.vscode/
/.idea/
/.projenrc.js
tsconfig.tsbuildinfo
/.eslintrc.json
!/assets/
cdk.out/
.cdk.staging/
/tsconfig.json
/tsconfig.dev.json
/.gitattributes
