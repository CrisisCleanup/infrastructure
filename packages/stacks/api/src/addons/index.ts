export * from './redis-stack'
export * from './crisiscleanup'
export * from './arc'
export * from './kube-prometheus-stack'
export * from './vertical-pod-autoscaler'
