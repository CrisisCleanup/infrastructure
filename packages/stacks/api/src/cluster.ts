import * as blueprints from '@aws-quickstart/eks-blueprints'
import { type CrisisCleanupConfig } from '@crisiscleanup/config'
import type * as ec2 from 'aws-cdk-lib/aws-ec2'
import { KubernetesVersion } from 'aws-cdk-lib/aws-eks'
import { KubectlLayer } from 'aws-cdk-lib/lambda-layer-kubectl'

enum Label {
	INSTANCE_TYPE = 'node.kubernetes.io/instance-type',
	TOPO_ZONE = 'topology.kubernetes.io/zone',
	ARCH = 'kubernetes.io/arch',
	CAPACITY_TYPE = 'karpenter.sh/capacity-type',
	KARPENTER_DISCOVERY = 'karpenter.sh/discovery',
	CLUSTER_DISCOVERY = 'kubernetes.io/cluster',
}

const defaultAddons: Array<blueprints.ClusterAddOn> = [
	// new blueprints.addons.ArgoCDAddOn(),
	// -- support k8s network policies
	// new blueprints.addons.CalicoOperatorAddOn(),
	// new blueprints.addons.ContainerInsightsAddOn(),
]

export const getDefaultAddons = (
	config: CrisisCleanupConfig,
): Array<blueprints.ClusterAddOn> => {
	const { apiStack } = config
	const { eks } = apiStack
	return [
		new blueprints.addons.MetricsServerAddOn(),
		new blueprints.addons.AwsLoadBalancerControllerAddOn(),
		new blueprints.addons.VpcCniAddOn({
			enablePrefixDelegation: true,
			version: eks.vpcCniVersion,
		}),
		new blueprints.addons.CoreDnsAddOn(eks.coreDnsVersion),
		new blueprints.addons.KubeProxyAddOn(eks.kubeProxyVersion),
	]
}

export const tagKarpenter = (stack: blueprints.EksBlueprint) => {
	const clusterInfo = stack.getClusterInfo()
	const vpc = clusterInfo.getResource<ec2.IVpc>(blueprints.GlobalResources.Vpc)!
	const discoveryTag = `${Label.KARPENTER_DISCOVERY}/${clusterInfo.cluster.clusterName}`
	blueprints.utils.tagSubnets(stack, vpc.privateSubnets, discoveryTag, '*')
}

export const buildKarpenter = (stackName: string, subnetNames: string[]) => {
	const clusterDiscoveryTag = `${Label.CLUSTER_DISCOVERY}/${stackName}`
	return new blueprints.KarpenterAddOn({
		version: 'v0.29.2',
		requirements: [
			{ key: Label.ARCH, op: 'In', vals: ['arm64'] },
			{ key: Label.CAPACITY_TYPE, op: 'In', vals: ['spot', 'on-demand'] },
		],
		subnetTags: {
			Name: subnetNames.join(','),
		},
		securityGroupTags: {
			[clusterDiscoveryTag]: 'owned',
		},
		amiFamily: 'AL2',
		consolidation: { enabled: true },
		interruptionHandling: true,
		namespace: 'karpenter',
		// refresh nodes at least every 30 days
		ttlSecondsUntilExpired: 2592000,
		// limits: {
		//   resources: {
		//     cpu: 10,
		//     memory: '200Gi',
		//   }
		// }
	})
}

export const buildEKSStack = (
	config: CrisisCleanupConfig,
): blueprints.BlueprintBuilder => {
	const { apiStack, cdkEnvironment } = config
	if (!apiStack) throw Error('No apistack config found.')
	return blueprints.EksBlueprint.builder()
		.account(String(cdkEnvironment.account))
		.region(cdkEnvironment.region)
		.version(KubernetesVersion.of(apiStack.eks.k8s.version))
		.addOns(...defaultAddons)
		.useDefaultSecretEncryption(apiStack.eks.defaultSecretsEncryption)
}

export const buildClusterBuilder = (
	config: CrisisCleanupConfig,
): blueprints.ClusterBuilder => {
	return blueprints.clusters
		.clusterBuilder()
		.withCommonOptions({
			version: KubernetesVersion.of(config.apiStack.eks.k8s.version),
		})
		.fargateProfile('serverless', { selectors: [{ namespace: 'karpenter' }] })
}
