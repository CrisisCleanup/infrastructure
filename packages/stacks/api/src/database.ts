import type * as blueprints from '@aws-quickstart/eks-blueprints'
import { type ResourceContext } from '@aws-quickstart/eks-blueprints'
import * as ec2 from 'aws-cdk-lib/aws-ec2'
import * as rds from 'aws-cdk-lib/aws-rds'
import { type ISecret, Secret } from 'aws-cdk-lib/aws-secretsmanager'

export interface DatabaseProviderProps {
	vpcResourceName: string
	databaseSecretResourceName: string

	engineVersion: string
	isolated: boolean
	ioOptimized: boolean

	databasePort?: number
}

export class DatabaseSecretProvider
	implements blueprints.ResourceProvider<ISecret>
{
	provide(context: ResourceContext): ISecret {
		return new Secret(context.scope, 'database-secret', {
			generateSecretString: {
				secretStringTemplate: JSON.stringify({ username: 'postgres' }),
				excludePunctuation: true,
				includeSpace: false,
				generateStringKey: 'password',
			},
		})
	}
}

export class DatabaseProvider
	implements blueprints.ResourceProvider<rds.DatabaseCluster>
{
	readonly props: DatabaseProviderProps

	constructor(props: DatabaseProviderProps) {
		this.props = props
	}

	provide(context: ResourceContext): rds.DatabaseCluster {
		const id = context.scope.node.id
		const databasePort = this.props.databasePort ?? 5432
		const subnetType = this.props.isolated
			? ec2.SubnetType.PRIVATE_ISOLATED
			: ec2.SubnetType.PRIVATE_WITH_EGRESS

		const credentialsSecret = context.get<ISecret>(
			this.props.databaseSecretResourceName,
		)
		if (!credentialsSecret) throw Error('Missing database credentials secret!')

		const vpc = context.get<ec2.IVpc>(this.props.vpcResourceName)
		if (!vpc) throw Error('Missing VPC!')

		const securityGroup = new ec2.SecurityGroup(
			context.scope,
			id + '-security-group',
			{
				vpc,
			},
		)

		securityGroup.addIngressRule(
			ec2.Peer.ipv4(vpc.vpcCidrBlock),
			ec2.Port.tcp(databasePort),
			'Ingress within VPC',
		)

		const engineVersion = rds.AuroraPostgresEngineVersion.of(
			this.props.engineVersion,
			this.props.engineVersion.split('.')[0],
			{ s3Export: true, s3Import: true },
		)
		const engine = rds.DatabaseClusterEngine.auroraPostgres({
			version: engineVersion,
		})

		const writer = rds.ClusterInstance.serverlessV2(id + '-cluster-writer')

		const clusterProps: rds.DatabaseClusterProps = {
			vpc,
			engine,
			vpcSubnets: {
				subnetType,
			},
			securityGroups: [securityGroup],
			credentials: rds.Credentials.fromSecret(credentialsSecret),
			iamAuthentication: true,
			port: databasePort,
			storageType: this.props.ioOptimized
				? rds.DBClusterStorageType.AURORA_IOPT1
				: rds.DBClusterStorageType.AURORA,
			storageEncrypted: true,
			serverlessV2MaxCapacity: 1,
			writer,
			readers: [rds.ClusterInstance.serverlessV2(id + '-cluster-reader-1')],
		}

		const cluster = new rds.DatabaseCluster(
			context.scope,
			id + '-database-cluster',
			clusterProps,
		)

		return cluster
	}
}
