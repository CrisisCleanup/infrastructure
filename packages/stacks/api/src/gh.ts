import assert from 'node:assert'
import * as blueprints from '@aws-quickstart/eks-blueprints'
import {
	GithubCodePipeline,
	type GithubWorkflowPipeline,
} from '@crisiscleanup/construct.awscdk.github-pipeline'
import type { Stack, StackProps } from 'aws-cdk-lib'
import * as cdk from 'aws-cdk-lib'
import * as kms from 'aws-cdk-lib/aws-kms'
import * as s3 from 'aws-cdk-lib/aws-s3'
import * as ghpipelines from 'cdk-pipelines-github'
import type { Construct } from 'constructs'

export interface GithubPipelineProps {
	repos?: string[]
	rootDir?: string
	owner?: string
}

export interface GithubCodePipelineProps
	extends Omit<blueprints.PipelineProps, 'repository'>,
		GithubPipelineProps {}

export interface StackStage extends blueprints.StackStage {
	stageProps?: ghpipelines.AddGitHubStageOptions
}

export interface PipelineWave extends blueprints.PipelineWave {
	stages: StackStage[]
}

/**
 * Extension of {@link @aws-quickstart/eks-blueprints#CodePipeline} for integration with {@link cdk-pipelines-github}
 */
export class GithubCodePipelineBuilder extends blueprints.CodePipelineBuilder {
	protected githubProps: GithubPipelineProps

	constructor() {
		super()
		this.githubProps = {}
	}

	build(
		scope: Construct,
		id: string,
		stackProps?: StackProps,
	): GithubCodePipelineStack {
		// @ts-ignore
		const fullProps = this.props as unknown as GithubCodePipelineProps
		const mergedProps = { ...fullProps, ...this.githubProps }
		return new GithubCodePipelineStack(scope, mergedProps, id, stackProps)
	}

	protected formatRepo(owner: string, repoString: string): string {
		if (repoString.includes('/')) return repoString
		return [owner, repoString].join('/')
	}

	owner(owner: string): this {
		super.owner(owner)
		this.githubProps.owner = owner
		if (this.githubProps.repos) {
			this.githubProps.repos = this.githubProps.repos.map((repo) =>
				this.formatRepo(owner, repo),
			)
		}
		return this
	}

	githubWave(...waves: PipelineWave[]): this {
		super.wave(...(waves as blueprints.PipelineWave[]))
		return this
	}

	githubRepo(...repos: string[]): this {
		const current = this.githubProps.repos ?? []
		this.githubProps.repos = Array.from(new Set([...current, ...repos]))
		const owner = this.githubProps.owner
		if (owner) {
			this.githubProps.repos = this.githubProps.repos.map((repo) =>
				this.formatRepo(owner, repo),
			)
		}
		return this
	}

	rootDir(value: string): this {
		this.githubProps.rootDir = value
		return this
	}
}

/**
 * Adapted from {@link @aws-quickstart/eks-blueprints#CodePipelineStack} for integration with {@link cdk-pipelines-github}
 * @see https://github.com/aws-quickstart/cdk-eks-blueprints/blob/main/lib/pipelines/code-pipeline.ts
 */
export class GithubCodePipelineStack extends cdk.Stack {
	static builder(): GithubCodePipelineBuilder {
		return new GithubCodePipelineBuilder()
	}

	private asyncTask: Promise<GithubWorkflowPipeline>

	constructor(
		scope: Construct,
		pipelineProps: GithubCodePipelineProps,
		id: string,
		props?: StackProps,
	) {
		super(scope, id, props)

		const pipeline = PipelineBuilder.build(this, pipelineProps)

		const promises: Promise<ApplicationStage>[] = []

		for (const stage of pipelineProps.stages) {
			const appStage = new ApplicationStage(this, stage.id, stage.stackBuilder)
			promises.push(appStage.waitForAsyncTasks())
		}

		this.asyncTask = Promise.all(promises)
			.then((stages) => {
				let currentWave: ghpipelines.GitHubWave | undefined

				// eslint-disable-next-line @typescript-eslint/no-for-in-array
				for (const i in stages) {
					const stage = pipelineProps.stages[i]
					if (stage.waveId) {
						if (currentWave == null || currentWave.id != stage.waveId) {
							const waveProps = pipelineProps.waves.find(
								(wave) => wave.id === stage.waveId,
							)
							assert(
								waveProps,
								`Specified wave ${stage.waveId} is not found in the pipeline definition ${id}`,
							)
							currentWave = pipeline.addGitHubWave(stage.waveId, {
								...waveProps.props,
							})
						}
						currentWave.addStageWithGitHubOptions(stages[i], stage.stageProps)
					} else {
						pipeline.addStageWithGitHubOptions(stages[i], stage.stageProps)
					}
				}
			})
			.catch(console.error)
			.then(() => pipeline)
	}

	async waitForAsyncTasks(): Promise<GithubWorkflowPipeline> {
		const pipeline = await this.asyncTask
		return pipeline
	}
}

/**
 * Adaption of {@link @aws-quickstart/eks-blueprints#ApplicationStage} for integration with {@link cdk-pipelines-github}.
 * @see https://github.com/aws-quickstart/cdk-eks-blueprints/blob/main/lib/pipelines/code-pipeline.ts
 */
export class ApplicationStage extends ghpipelines.GitHubStage {
	private asyncTask: Promise<any> | undefined = undefined

	constructor(
		scope: Stack,
		id: string,
		builder: blueprints.StackBuilder | blueprints.AsyncStackBuilder,
		props?: ghpipelines.GitHubStageProps,
	) {
		super(scope, id, props)
		if ((<blueprints.AsyncStackBuilder>builder).buildAsync !== undefined) {
			this.asyncTask = (<blueprints.AsyncStackBuilder>builder).buildAsync(
				this,
				`${id}-blueprint`,
				props,
			)
		} else {
			builder.build(this, `${id}-blueprint`, props)
		}
	}

	public async waitForAsyncTasks(): Promise<ApplicationStage> {
		if (this.asyncTask) {
			return this.asyncTask.then(() => {
				return this
			})
		}
		return Promise.resolve(this)
	}
}

class PipelineBuilder {
	static build(scope: Construct, props: GithubCodePipelineProps) {
		const actionsRole = new ghpipelines.GitHubActionRole(
			scope,
			'github-action-role',
			{
				repos: props.repos ?? [],
			},
		)

		const pipelineS3BucketName = 'crisiscleanup-pipeline-assets'
		const pipelineS3 = new s3.Bucket(scope, 'pipeline-assets', {
			autoDeleteObjects: true,
			removalPolicy: cdk.RemovalPolicy.DESTROY,
			transferAcceleration: true,
			blockPublicAccess: s3.BlockPublicAccess.BLOCK_ALL,
			bucketName: pipelineS3BucketName,
		})
		pipelineS3.grantReadWrite(actionsRole.role)
		pipelineS3.addLifecycleRule({
			id: 'cleanup-stale-assets',
			enabled: true,
			expiration: cdk.Duration.days(30),
			abortIncompleteMultipartUploadAfter: cdk.Duration.days(7),
			prefix: 'cdk-assets',
		})
		pipelineS3.addLifecycleRule({
			id: 'cleanup-build-stale-cache',
			enabled: true,
			expiration: cdk.Duration.days(30),
			abortIncompleteMultipartUploadAfter: cdk.Duration.days(7),
			prefix: 'build-cache',
		})

		const pipelineKms = new kms.Key(scope, 'pipeline-kms', {
			alias: 'pipeline-key',
		})
		pipelineKms.grantDecrypt(actionsRole.role)

		return GithubCodePipeline.create({
			rootDir: props.rootDir!,
			workflowName: 'Deploy',
			assetsS3Bucket: pipelineS3BucketName,
		})
			.addNxEnv()
			.addConfigsEnv()
			.synthTarget({
				packageName: 'stacks.api',
				workingDirectory: 'packages/stacks/api',
			})
			.defaultTools()
			.clone({ workflowTriggers: {} })
			.build(scope, props.name)
			.onWorkflowCall({
				runner: {
					type: 'string',
					default: 'ubuntu-latest',
					description: 'Runner to use.',
					required: false,
				},
				environments: {
					type: 'string',
					description: 'Environments to deploy.',
					default: 'development,staging',
					required: false,
				},
			})
			.onWorkflowDispatch({
				runner: {
					type: 'choice',
					description: 'Runner to use.',
					options: ['ubuntu-latest', 'self-hosted'],
					default: 'ubuntu-latest',
				},
				environments: {
					type: 'choice',
					description: 'Environments to deploy.',
					options: [
						'development',
						'staging',
						'production',
						'development,staging',
						'development,staging,production',
					],
					default: 'development,staging',
				},
			})
			.concurrency({
				group: 'deploy-infra',
				cancelInProgress: false,
			})
	}
}
