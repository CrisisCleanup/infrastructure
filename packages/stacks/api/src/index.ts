export * from './schema'
export * from './addons'
export * from './stacks'
export * from './cluster'
export * from './gh'
export * from './secrets'
export * from './kms'
export * from './util'
export * from './pipeline'
