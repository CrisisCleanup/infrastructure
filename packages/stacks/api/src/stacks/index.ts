export * from './network'
export * from './data'
export * from './cache'
export * from './zones'
