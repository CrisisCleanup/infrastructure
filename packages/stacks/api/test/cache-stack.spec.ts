import { App, Stack } from 'aws-cdk-lib'
import { Template } from 'aws-cdk-lib/assertions'
import { Vpc } from 'aws-cdk-lib/aws-ec2'
import { beforeEach, describe, expect, test } from 'vitest'
import { cacheConfigSchema } from '../src'
import { CacheStack } from '../src/stacks'

interface TestContext {
	app: App
	vpc: Vpc
}

describe('CacheStack', () => {
	beforeEach<TestContext>((ctx) => {
		ctx.app = new App()
		const vpcStack = new Stack(ctx.app, 'test-vpc')
		ctx.vpc = new Vpc(vpcStack, 'test-vpc')
	})

	test<TestContext>('renders expected template with defaults', (ctx) => {
		const cacheProps = cacheConfigSchema.parse({
			enabled: true,
		})
		const stack = new CacheStack(ctx.app, 'test-cache', {
			vpc: ctx.vpc,
			...cacheProps,
		})
		const template = Template.fromStack(stack)
		expect(template.toJSON()).toMatchSnapshot()
	})

	test<TestContext>('renders expected template with cluster mode', (ctx) => {
		const cacheProps = cacheConfigSchema.parse({
			enabled: true,
			clusterMode: true,
			replicas: 3,
			memoryAutoscalingTarget: 60,
		})
		const stack = new CacheStack(ctx.app, 'test-cache', {
			vpc: ctx.vpc,
			...cacheProps,
		})
		const template = Template.fromStack(stack)
		expect(template.toJSON()).toMatchSnapshot()
	})
})
