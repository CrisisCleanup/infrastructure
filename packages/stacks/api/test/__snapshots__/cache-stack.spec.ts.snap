// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`CacheStack > renders expected template with cluster mode 1`] = `
{
  "Parameters": {
    "BootstrapVersion": {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "testcacheredisScalableTarget51BB6BE3": {
      "Properties": {
        "MaxCapacity": 1,
        "MinCapacity": 1,
        "ResourceId": {
          "Fn::Join": [
            "",
            [
              "replication-group/",
              {
                "Ref": "testcacheredistestcacheredisRedisDBB8E82A88",
              },
            ],
          ],
        },
        "RoleARN": {
          "Fn::GetAtt": [
            "testcacheredisScalableTargetRoleA095D194",
            "Arn",
          ],
        },
        "ScalableDimension": "elasticache:replication-group:NodeGroups",
        "ServiceNamespace": "elasticache",
      },
      "Type": "AWS::ApplicationAutoScaling::ScalableTarget",
    },
    "testcacheredisScalableTargetMemTracking667E3C0F": {
      "Properties": {
        "PolicyName": "testcachetestcacheredisScalableTargetMemTracking8543E3D8",
        "PolicyType": "TargetTrackingScaling",
        "ScalingTargetId": {
          "Ref": "testcacheredisScalableTarget51BB6BE3",
        },
        "TargetTrackingScalingPolicyConfiguration": {
          "PredefinedMetricSpecification": {
            "PredefinedMetricType": "ElastiCacheDatabaseMemoryUsageCountedForEvictPercentage",
          },
          "TargetValue": 60,
        },
      },
      "Type": "AWS::ApplicationAutoScaling::ScalingPolicy",
    },
    "testcacheredisScalableTargetRoleA095D194": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "application-autoscaling.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::IAM::Role",
    },
    "testcacheredistestcacheredisRedisDBB8E82A88": {
      "DependsOn": [
        "testcacheredistestcacheredisRedisDBSubnetGroup62583AC4",
      ],
      "Properties": {
        "AtRestEncryptionEnabled": true,
        "AutoMinorVersionUpgrade": false,
        "CacheNodeType": "cache.r7g.xlarge",
        "CacheParameterGroupName": "default.redis7.cluster.on",
        "CacheSubnetGroupName": "RedisDBSubnetGroup",
        "Engine": "Redis",
        "EngineVersion": "7.1",
        "MultiAZEnabled": false,
        "NumNodeGroups": 1,
        "ReplicasPerNodeGroup": 3,
        "ReplicationGroupDescription": "RedisDB setup by CDK",
        "SecurityGroupIds": [
          {
            "Fn::GetAtt": [
              "testcachesecuritygroup6F4AC382",
              "GroupId",
            ],
          },
        ],
      },
      "Type": "AWS::ElastiCache::ReplicationGroup",
    },
    "testcacheredistestcacheredisRedisDBSubnetGroup62583AC4": {
      "Properties": {
        "CacheSubnetGroupName": "RedisDBSubnetGroup",
        "Description": "RedisDB Subnet Group",
        "SubnetIds": [
          {
            "Fn::ImportValue": "test-vpc:ExportsOutputReftestvpcPrivateSubnet1Subnet865FB50A19DF49CF",
          },
          {
            "Fn::ImportValue": "test-vpc:ExportsOutputReftestvpcPrivateSubnet2Subnet23D3396F45FDAD97",
          },
        ],
      },
      "Type": "AWS::ElastiCache::SubnetGroup",
    },
    "testcachesecuritygroup6F4AC382": {
      "Properties": {
        "GroupDescription": "Security group for RedisDB",
        "SecurityGroupEgress": [
          {
            "CidrIp": {
              "Fn::ImportValue": "test-vpc:ExportsOutputFnGetAtttestvpc8985080ECidrBlockEB7C13CD",
            },
            "Description": "Egress within VPC",
            "FromPort": 0,
            "IpProtocol": "tcp",
            "ToPort": 65535,
          },
        ],
        "SecurityGroupIngress": [
          {
            "CidrIp": {
              "Fn::ImportValue": "test-vpc:ExportsOutputFnGetAtttestvpc8985080ECidrBlockEB7C13CD",
            },
            "Description": "Ingress within VPC",
            "FromPort": 6379,
            "IpProtocol": "tcp",
            "ToPort": 6379,
          },
        ],
        "VpcId": {
          "Fn::ImportValue": "test-vpc:ExportsOutputReftestvpc8985080E5120E245",
        },
      },
      "Type": "AWS::EC2::SecurityGroup",
    },
  },
  "Rules": {
    "CheckBootstrapVersion": {
      "Assertions": [
        {
          "Assert": {
            "Fn::Not": [
              {
                "Fn::Contains": [
                  [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;

exports[`CacheStack > renders expected template with defaults 1`] = `
{
  "Parameters": {
    "BootstrapVersion": {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "testcacheredistestcacheredisRedisDBB8E82A88": {
      "DependsOn": [
        "testcacheredistestcacheredisRedisDBSubnetGroup62583AC4",
      ],
      "Properties": {
        "AtRestEncryptionEnabled": true,
        "AutoMinorVersionUpgrade": false,
        "CacheNodeType": "cache.r7g.xlarge",
        "CacheParameterGroupName": "default.redis7",
        "CacheSubnetGroupName": "RedisDBSubnetGroup",
        "Engine": "Redis",
        "EngineVersion": "7.1",
        "MultiAZEnabled": false,
        "NumNodeGroups": 1,
        "ReplicasPerNodeGroup": 1,
        "ReplicationGroupDescription": "RedisDB setup by CDK",
        "SecurityGroupIds": [
          {
            "Fn::GetAtt": [
              "testcachesecuritygroup6F4AC382",
              "GroupId",
            ],
          },
        ],
      },
      "Type": "AWS::ElastiCache::ReplicationGroup",
    },
    "testcacheredistestcacheredisRedisDBSubnetGroup62583AC4": {
      "Properties": {
        "CacheSubnetGroupName": "RedisDBSubnetGroup",
        "Description": "RedisDB Subnet Group",
        "SubnetIds": [
          {
            "Fn::ImportValue": "test-vpc:ExportsOutputReftestvpcPrivateSubnet1Subnet865FB50A19DF49CF",
          },
          {
            "Fn::ImportValue": "test-vpc:ExportsOutputReftestvpcPrivateSubnet2Subnet23D3396F45FDAD97",
          },
        ],
      },
      "Type": "AWS::ElastiCache::SubnetGroup",
    },
    "testcachesecuritygroup6F4AC382": {
      "Properties": {
        "GroupDescription": "Security group for RedisDB",
        "SecurityGroupEgress": [
          {
            "CidrIp": {
              "Fn::ImportValue": "test-vpc:ExportsOutputFnGetAtttestvpc8985080ECidrBlockEB7C13CD",
            },
            "Description": "Egress within VPC",
            "FromPort": 0,
            "IpProtocol": "tcp",
            "ToPort": 65535,
          },
        ],
        "SecurityGroupIngress": [
          {
            "CidrIp": {
              "Fn::ImportValue": "test-vpc:ExportsOutputFnGetAtttestvpc8985080ECidrBlockEB7C13CD",
            },
            "Description": "Ingress within VPC",
            "FromPort": 6379,
            "IpProtocol": "tcp",
            "ToPort": 6379,
          },
        ],
        "VpcId": {
          "Fn::ImportValue": "test-vpc:ExportsOutputReftestvpc8985080E5120E245",
        },
      },
      "Type": "AWS::EC2::SecurityGroup",
    },
  },
  "Rules": {
    "CheckBootstrapVersion": {
      "Assertions": [
        {
          "Assert": {
            "Fn::Not": [
              {
                "Fn::Contains": [
                  [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
